### ISSUE TITLE
[AMDGPU] llc failed at O1 and O3 but works on O0 and O2

### ISSUE THREAD

**Issue by chichunchen** [llvm/llvm-project#54157]
https://godbolt.org/z/M4nfWenhM

Command:
```
llc -O3 -mtriple=amdgcn-amd-amdhsa -disable-promote-alloca-to-lds -mcpu=gfx906 -filetype=asm bugpoint-reduced-simplified.ll -o test.s
```

IR:
```
; ModuleID = '<stdin>'
source_filename = "llvm-link"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7"
target triple = "amdgcn-amd-amdhsa"

%DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 = type { double*, %dv_hdr.0.4.8.28.32.36.44.52.56.60.64.68.76.96.144, %dv_dim.1.5.9.29.33.37.45.53.57.61.65.69.77.97.145 }
%dv_hdr.0.4.8.28.32.36.44.52.56.60.64.68.76.96.144 = type { i64, i56, i40, i32, i64, i64 }
%dv_dim.1.5.9.29.33.37.45.53.57.61.65.69.77.97.145 = type { i64, i64, i64 }

define dso_local void @"loginterpolatesinglevariable_2d_custom_point$test_mod_"() #0 !scalarlevel !1 !cachelevel !1 !fplevel !1 {
", bb1":
  %"t$6" = alloca %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146, align 16, addrspace(5)
  %r57 = addrspacecast %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 addrspace(5)* %"t$6" to i8*
  %r58 = bitcast i8* %r57 to %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146*
  %r9.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %r58 to i64
  %r10.i = inttoptr i64 %r9.i to i64*
  %r68.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %r58 to i64
  %r69.i = inttoptr i64 %r68.i to i64*
  %r70.i = getelementptr i64, i64* %r69.i, i64 6
  ret void
}

attributes #0 = { "target-cpu"="gfx906" }

!opencl.ocl.version = !{!0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0}

!0 = !{i32 2, i32 0}
!1 = !{i64 0}
```

Error message:
```
Instruction does not dominate all uses!
  %1 = addrspacecast i64 addrspace(5)* %0 to %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146*
  %r9.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %1 to i64
in function loginterpolatesinglevariable_2d_custom_point$test_mod_
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llc -O3 -mtriple=amdgcn-amd-amdhsa -disable-promote-alloca-to-lds -mcpu=gfx906 -filetype=asm bugpoint-reduced-simplified.ll -o test.s
1.      Running pass 'Function Pass Manager' on module 'bugpoint-reduced-simplified.ll'.
2.      Running pass 'Module Verifier' on function '@"loginterpolatesinglevariable_2d_custom_point$test_mod_"'
 #0 0x000000000566b3ca llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /ptmp/cchen/llvm-project/llvm/lib/Support/Unix/Signals.inc:565:11
 #1 0x000000000566b57b PrintStackTraceSignalHandler(void*) /ptmp/cchen/llvm-project/llvm/lib/Support/Unix/Signals.inc:632:1
 #2 0x0000000005669c2a llvm::sys::RunSignalHandlers() /ptmp/cchen/llvm-project/llvm/lib/Support/Signals.cpp:97:5
 #3 0x000000000566bca5 SignalHandler(int) /ptmp/cchen/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007fdb8cfe5b20 __restore_rt sigaction.c:0:0
 #5 0x00007fdb8bcf337f raise (/lib64/libc.so.6+0x3737f)
 #6 0x00007fdb8bcdddb5 abort (/lib64/libc.so.6+0x21db5)
 #7 0x000000000558ad4a (/ptmp/cchen/myllvm/bin/llc+0x558ad4a)
 #8 0x000000000558abc2 /ptmp/cchen/llvm-project/llvm/lib/Support/ErrorHandling.cpp:83:3
 #9 0x0000000004b1e7bb (anonymous namespace)::VerifierLegacyPass::runOnFunction(llvm::Function&) /ptmp/cchen/llvm-project/llvm/lib/IR/Verifier.cpp:6021:5
#10 0x0000000004a2672e llvm::FPPassManager::runOnFunction(llvm::Function&) /ptmp/cchen/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1434:23
#11 0x0000000004a2b5b2 llvm::FPPassManager::runOnModule(llvm::Module&) /ptmp/cchen/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1480:16
#12 0x0000000004a27019 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /ptmp/cchen/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1549:23
#13 0x0000000004a26b8d llvm::legacy::PassManagerImpl::run(llvm::Module&) /ptmp/cchen/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:16
#14 0x0000000004a2b891 llvm::legacy::PassManager::run(llvm::Module&) /ptmp/cchen/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1676:3
#15 0x0000000003016cfc compileModule(char**, llvm::LLVMContext&) /ptmp/cchen/llvm-project/llvm/tools/llc/llc.cpp:722:41
#16 0x000000000301509e main /ptmp/cchen/llvm-project/llvm/tools/llc/llc.cpp:417:13
#17 0x00007fdb8bcdf493 __libc_start_main (/lib64/libc.so.6+0x23493)
#18 0x000000000301489e _start (/ptmp/cchen/myllvm/bin/llc+0x301489e)
Aborted (core dumped)
```

**Comment by chichunchen** [comment]
The issue happens when there's two pairs of ptrtoint/inttoptr, however, if we remove one pair of ptrtoint/inttoptr, then we can avoid the error from the verifier.

```
%DV1 = type { double*, %dv_hdr, %dv_dim }
%dv_hdr = type { i64, i56, i40, i32, i64, i64 }
%dv_dim = type { i64, i64, i64 }

define dso_local void @"loginterpolatesinglevariable_2d_custom_point$test_mod_"() #0 !scalarlevel !1 !cachelevel !1 !fplevel !1 {
", bb1":
  %"t$6" = alloca %DV1, align 16, addrspace(5)
  %r57 = addrspacecast %DV1 addrspace(5)* %"t$6" to i8*
  %r58 = bitcast i8* %r57 to %DV1*
  %r9.i = ptrtoint %DV1* %r58 to i64
  %r10.i = inttoptr i64 %r9.i to i64*
  %r68.i = ptrtoint %DV1* %r58 to i64             <-- remove
  %r69.i = inttoptr i64 %r68.i to i64*                <-- remove
  %r70.i = getelementptr i64, i64* %r69.i, i64 0
  ret void
}
```

Before: https://godbolt.org/z/3xE6Ybo35
After: https://godbolt.org/z/Tfe3Kd9no

**Comment by jayfoad** [comment]
infer-address-spaces is changing %r9.i to refer to a new instruction %1 which is inserted _after_ %r9.i.

Before infer-address-spaces:
```
define dso_local void @"loginterpolatesinglevariable_2d_custom_point$test_mod_"() #0 !scalarlevel !1 !cachelevel !1 !fplevel !1 {
", bb1":
  %"t$6" = alloca %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146, align 16, addrspace(5)
  %r57 = addrspacecast %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 addrspace(5)* %"t$6" to i8*
  %r58 = bitcast i8* %r57 to %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146*
  %r9.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %r58 to i64
  %r10.i = inttoptr i64 %r9.i to i64*
  %r68.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %r58 to i64
  %r69.i = inttoptr i64 %r68.i to i64*
  %r70.i = getelementptr i64, i64* %r69.i, i64 6
  ret void
}
```
After infer-address-spaces:
```
define dso_local void @"loginterpolatesinglevariable_2d_custom_point$test_mod_"() #0 !scalarlevel !1 !cachelevel !1 !fplevel !1 {
", bb1":
  %"t$6" = alloca %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146, align 16, addrspace(5)
  %r57 = bitcast %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 addrspace(5)* %"t$6" to i8 addrspace(5)*
  %r58 = bitcast i8 addrspace(5)* %r57 to %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 addrspace(5)*
  %r9.i = ptrtoint %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146* %1 to i64
  %r10.i = inttoptr i64 %r9.i to i64*
  %0 = bitcast %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146 addrspace(5)* %r58 to i64 addrspace(5)*
  %1 = addrspacecast i64 addrspace(5)* %0 to %DV1.2.6.10.30.34.38.46.54.58.62.66.70.78.98.146*
  %2 = addrspacecast i64 addrspace(5)* %0 to i64*
  %r70.i = getelementptr i64, i64* %2, i64 6
  ret void
}
```
@kerbowa maybe related to https://reviews.llvm.org/D119524?

**Comment by kerbowa** [comment]
I'll take a look. Doesn't seem like reverting [D119524](https://reviews.llvm.org/D119524) is helping.

**Comment by chichunchen** [comment]
Any updates about this one?

**Comment by kerbowa** [comment]
Should be fixed by 26b14c3ea7

**Comment by fhahn** [comment]
Looks fixed on latest `main`: https://llvm.godbolt.org/z/G3799TMqT

Please re-open if the crash still reproduces on your end.
